core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -141,41 +141,43 @@ private void removeRecovery(ShardRouting routing) {
         addRecovery(routing, false, null);
     }
 
-    private void addInitialRecovery(ShardRouting routing, RoutingTable routingTable) {
-        addRecovery(routing, true, routingTable);
+    private void addInitialRecovery(ShardRouting routing, ShardRouting initialPrimaryShard) {
+        addRecovery(routing, true, initialPrimaryShard);
     }
 
-    private void addRecovery(final ShardRouting routing, final boolean increment, final RoutingTable routingTable) {
+    private void addRecovery(final ShardRouting routing, final boolean increment, final ShardRouting initialPrimaryShard) {
         final int howMany = increment ? 1 : -1;
         assert routing.initializing() : "routing must be initializing: " + routing;
+        assert initialPrimaryShard == null || initialPrimaryShard.assignedToNode() :
```

Comment:
I had that one originally there. But this fails some tests currently as we don't always add the ReplicaAfterPrimaryActiveAllocationDecider to the list of deciders in these tests. As this impacts more than a few places, I decided against tackling that one in this PR.